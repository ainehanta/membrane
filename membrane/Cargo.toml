[package]
authors = ["Jerel Unruh <mail@unruhdesigns.com>"]
description = "Membrane is an opinionated crate that generates a Dart package from a Rust library. Extremely fast performance with strict typing and zero copy returns over the FFI boundary via bincode."
edition = "2018"
keywords = ["dart", "ffi", "flutter", "bindings", "codegen"]
license = "Apache-2.0"
name = "membrane"
readme = "../README.md"
repository = "https://github.com/jerel/membrane"
version = "0.6.2"

[lib]
crate-type = ["lib"]
doctest = false

[features]
generate = ["membrane_macro/generate"]
skip-generate = ["membrane_macro/skip-generate"]

[dependencies]
allo-isolate = "0.1.11"
bincode = {version = "1.3"}
ffi_helpers = "0.2"
futures = "0.3"
heck = "0.3"
inventory = "0.1"
membrane_macro = {version = "^0.5", path = "../membrane_macro"}
membrane_types = {version = "^0.3", path = "../membrane_types"}
regex = "1.5"
serde = {version = "1.0", features = ["derive"]}
serde-generate = "0.20.4"
serde-reflection = "0.3.5"

[dev-dependencies]
example = {path = "../example"}
pretty_assertions = "1.0.0"
serial_test = "0.5.1"
trybuild = "1.0"
