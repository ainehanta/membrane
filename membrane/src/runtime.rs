use once_cell::sync::OnceCell;
use std::{fmt::Debug, future::Future};

pub trait Interface {
  fn spawn<F>(&self, future: F) -> JoinHandle
  where
    F: Future + Send + 'static,
    F::Output: Send + Debug + 'static;

  fn spawn_blocking<F, R>(&self, future: F) -> JoinHandle
  where
    F: FnOnce() -> R + Send + 'static,
    R: Send + Debug + 'static;
}

pub struct JoinHandle {
  pub debug_id: String,
  pub abort: Box<dyn Fn()>,
}

impl JoinHandle {
  pub fn abort(&self) {
    (self.abort)();
  }
}

impl Debug for JoinHandle {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    f.debug_struct("JoinHandle")
      .field("debug_id", &self.debug_id)
      .finish()
  }
}

#[derive(Debug)]
pub struct App<T: Interface> {
  builder: fn() -> T,
  runtime: OnceCell<T>,
  metadata: Vec<i32>,
}

impl<T: Interface> App<T> {
  pub const fn new(builder: fn() -> T) -> App<T> {
    App {
      builder,
      runtime: OnceCell::new(),
      metadata: vec![],
    }
  }

  pub fn get(&self) -> &T {
    self.runtime.get_or_init(self.builder)
  }

  pub fn metadata(&self) -> &Vec<i32> {
    &self.metadata
  }
}
