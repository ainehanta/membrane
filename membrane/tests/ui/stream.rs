use futures::Stream;
use membrane::async_dart;
use membrane::runtime::{App, Interface, JoinHandle};
use std::{fmt::Debug, future::Future};

struct TestRuntime();
impl Interface for TestRuntime {
  fn spawn<T>(&self, future: T) -> JoinHandle
  where
    T: Future + Send + 'static,
    T::Output: Send + 'static,
  {
    JoinHandle {
      debug_id: String::new(),
      abort: Box::new(|| {}),
    }
  }

  fn spawn_blocking<F, R>(&self, future: F) -> JoinHandle
  where
    F: FnOnce() -> R + Send + 'static,
    R: Send + Debug + 'static,
  {
    JoinHandle {
      debug_id: String::new(),
      abort: Box::new(|| {}),
    }
  }
}

static RUNTIME: App<TestRuntime> = App::new(|| TestRuntime());

#[async_dart(namespace = "a")]
pub fn one_failure() -> impl Stream<i32, String> {}

#[async_dart(namespace = "a")]
pub fn two_failure() -> impl Stream<Item = i32, String> {}

#[async_dart(namespace = "a")]
pub fn one_success() -> impl Stream<Item = Result<i32, String>> {
  futures::stream::iter(vec![])
}

fn main() {}
