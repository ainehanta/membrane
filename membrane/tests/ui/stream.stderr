error: expected `impl Stream<Item = Result>`
  --> tests/ui/stream.rs:34:30
   |
34 | pub fn one_failure() -> impl Stream<i32, String> {}
   |                              ^^^^^^

error: expected enum `Result`
  --> tests/ui/stream.rs:37:44
   |
37 | pub fn two_failure() -> impl Stream<Item = i32, String> {}
   |                                            ^^^
